data-processing/common/upload_entities.py

```diff
@@ -155,73 +109,193 @@ def upload_entities(
+        data_version = get_or_create_data_version(s2_id)
+
+    entity_models: EntityModels = {}
     uploaded_entities = Entity.select().where(
         Entity.paper_id == s2_id, Entity.version == data_version
     )
     for entity_model in uploaded_entities:
-        uploaded_entity_models[
-            (entity_model.type, entity_model.within_paper_id)
-        ] = entity_model
+        entity_models[(entity_model.type, entity_model.within_paper_id)] = entity_model
+
+    return entity_models
+
+
+def make_data_models(
+    entity_model: Optional[Entity], entity_model_id: Optional[str], data: EntityData
+) -> List[EntityDataModel]:
+    " Either 'entity_model' or 'entity_model_id' must be defined. "
+
+    if entity_model is None and entity_model_id is None:
+        logging.warning(  # pylint: disable=logging-not-lazy
+            "Attempted to create data models without providing an entity model or explicit ID. "
+            + "At least one of these must be provided to determine which entity the data should "
+            + "be associated with."
+        )
+        return []
+
+    models: List[EntityDataModel] = []
 
-    def resolve_model(entity_type: str, within_paper_id: str) -> Optional[Entity]:
+    # Create models for each field in the entity data.
+    for key, value in data.items():
+
+        # The value may have been specified as a list, or as a single scalar value.
+        # Unpack all of the values for this key into a list.
+        values: List[Any] = []
+        if isinstance(value, list):
+            of_list = True
+            values.extend(value)
+        else:
+            of_list = False
+            values.append(value)
+
+        if len(values) == 0:
+            continue
+
+        value_types = {type(v) for v in values}
+        if not len(value_types) == 1:
+            logging.warning(  # pylint: disable=logging-not-lazy
+                "Attempted to upload multiple primitive types of data for key %s. "
+                + "Types were %s. Not permitted. Skipping this value.",
+                key,
+                value_types,
+            )
+            continue
+
+        # Create a new row for each value, with information of the base data type
+        # and whether that row belongs to a list. If casting of values needs to occur
+        # for values based on type to make them appropriate for insertion in Postgres
+        # (e.g., casting booleans to 0 / 1), that should happen here.
+        for v in values:
+            type_ = None
+            # Check for boolean needs to come before check for int, because booleans
+            # will pass the check 'isinstance(v, int)'.
+            if isinstance(v, bool):
+                type_ = "boolean"
+                v = 1 if v else 0
+            if isinstance(v, int):
+                type_ = "integer"
+            elif isinstance(v, float):
+                type_ = "float"
+            elif isinstance(v, str):
+                type_ = "string"
```

Comment:
Missing some sort of else condition if the type is not one of these? I see that you check None-ness below, so perhaps a DEBUG log (or INFO-level) of some sort? DEBUG makes sense to me b/c the assumption here is that wherever we've generated these values/types can only provide these, however, in the long-term it's easy to forget that constraint!

Reply:
Good thinking! I'll add this statement.